pkg/kademlia/kademlia.go

```diff
@@ -5,243 +5,198 @@ package kademlia
 func (k *Kademlia) GetRoutingTable(ctx context.Context) (dht.RoutingTable, error) {
-	return &RoutingTable{
-		// ht:  k.dht.HT,
-		// dht: k.dht,
-	}, nil
+	return k.routingTable, nil
 
 }
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k *Kademlia) Bootstrap(ctx context.Context) error {
-	return k.dht.Bootstrap()
-}
-
-// Ping checks that the provided node is still accessible on the network
-func (k *Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
-	n, err := convertProtoNode(node)
-	if err != nil {
-		return proto.Node{}, err
+	// What I want to do here is do a normal lookup for myself
+	// so call lookup(ctx, nodeImLookingFor)
+	if len(k.bootstrapNodes) == 0 {
+		return BootstrapErr.New("no bootstrap nodes provided")
 	}
 
-	ok, err := k.dht.Ping(n)
-	if err != nil {
-		return proto.Node{}, err
-	}
-	if !ok {
-		return proto.Node{}, NodeErr.New("node unavailable")
-	}
-	return node, nil
+	_, err := k.lookup(ctx, node.StringToID(k.routingTable.self.GetId()), lookupOpts{amount: 5})
+	return err
 }
 
-// FindNode looks up the provided NodeID first in the local Node, and if it is not found
-// begins searching the network for the NodeID. Returns and error if node was not found
-func (k *Kademlia) FindNode(ctx context.Context, ID dht.NodeID) (proto.Node, error) {
-	nodes, err := k.dht.FindNode(ID.Bytes())
-	if err != nil {
-		return proto.Node{}, err
-
+func (k *Kademlia) lookup(ctx context.Context, target dht.NodeID, opts lookupOpts) (*proto.Node, error) {
+	conc := opts.amount
+	kb := k.routingTable.K()
+	if conc <= 0 {
+		conc = kb
 	}
 
-	for _, v := range nodes {
-		if string(v.ID) == ID.String() {
-			return proto.Node{Id: string(v.ID), Address: &proto.NodeAddress{
-				Transport: defaultTransport,
-				Address:   fmt.Sprintf("%s:%d", v.IP.String(), v.Port),
-			},
-			}, nil
-		}
+	// look in routing table for targetID
+	nodes, err := k.routingTable.FindNear(target, kb)
+	if err != nil {
+		return nil, err
 	}
-	return proto.Node{}, NodeErr.New("node not found")
-}
 
-// ListenAndServe connects the kademlia node to the network and listens for incoming requests
-func (k *Kademlia) ListenAndServe() error {
-	if err := k.dht.CreateSocket(); err != nil {
-		return err
+	// if we have the node in our routing table and that node is not this node (bootstrap) just return the node
+	if len(nodes) == 1 {
+		return nodes[0], nil
 	}
 
-	go func() {
-		if err := k.dht.Listen(); err != nil {
-			log.Printf("Failed to listen on the dht: %s\n", err)
-		}
-	}()
-
-	return nil
-}
+	// begin the work looking for the node by spinning up alpha workers
+	// and asking for the node we want from nodes we know in our routing table
+	ch := make(chan []*proto.Node, conc)
+	w := newWorker(ctx, k.routingTable, nodes, k.nodeClient, target, kb)
+	ctx, w.cancel = context.WithCancel(ctx)
+	defer w.cancel()
+	for i := 0; i < conc; i++ {
```

Comment:
oh, alpha actually should be the concurrency, not k. but each worker should request k at a time

Reply:
haha I swear I was using alpha originally. I'll fix.